### PR TITLE
[REVIEW] cudf::allocate_like can force mask allocation

### PR DESCRIPTION
**Pull request by cwharris**
Giving allocate_like the ability to allocate a mask regardless of the original table/column, such that it can be used when we know the new table will require a mask, but the original does not have one. An example of this is `shift`, where the original table may be completely valid, but the fill_value for the shift may be invalid, therefore requiring the newly allocated table to have a mask.

### CODE REVIEW

**Comment by rgsl888prabhu** [comment]
@harrism @cwharris should we also take care of initializing the mask ?

**Comment by jrhemstad** [comment]
> @harrism @cwharris should we also take care of initializing the mask ?

`allocate_like` is intended only to allocate uninitialized memory.

File: cpp/include/cudf/copying.hpp [hunk 1]
```diff
@@ -26,8 +26,8 @@ typedef struct CUstream_st* cudaStream_t;
 
 namespace cudf {
 
-enum MaskAlloc {
-    RETAIN,
+enum mask_allocation_policy {
+    RETAIN, ///< Allocate a null mask if the input contains one
     NEVER,
```
Comment (harrism) [inline cpp/include/cudf/copying.hpp]: Document every enum value.

File: cpp/include/cudf/copying.hpp [hunk 2]
```diff
@@ -44,22 +44,24 @@ gdf_column empty_like(gdf_column const& input);
  * @brief Allocates a new column of the same size and type as the input.
  *
  * @param input The input column to emulate
- * @param mask_alloc Determines under what condition to allocate the mask. Defaults to RETAIN.
+ * @param mask_alloc Policy for allocating null mask. Defaults to RETAIN.
  * @param stream Optional stream in which to perform copies
  * @return gdf_column An allocated column of same size and type of input
  */
-gdf_column allocate_like(gdf_column const& input, MaskAlloc mask_alloc = MaskAlloc::RETAIN, cudaStream_t stream = 0);
+gdf_column allocate_like(gdf_column const& input, mask_allocation_policy mask_alloc = mask_allocation_policy::RETAIN,
```
Comment (harrism) [inline cpp/include/cudf/copying.hpp]: No need to qualify every RETAIN with `mask_allocation_policy::`. it makes lines like this one too long (at least wrap the line), and it turns a lot of would-be one-liners into multi-line expressions.

Comment (cwharris) [reply cpp/include/cudf/copying.hpp]: I was unaware. The compiler infers the type? Or am I missing something...

Comment (harrism) [reply cpp/include/cudf/copying.hpp]: This is an unscoped enum, which puts its values into the surrounding scope (`namespace cudf` in this case). C++11 added `enum class`, which requires the kind of scoping you are using. However Cython doesn't support it yet, so we can't use it in cuDF public APIs like this one.

Comment (harrism) [reply cpp/include/cudf/copying.hpp]: To answer your question, I don't think there's a type to infer for an unscoped enum. its values just have integral type.

**Comment by harrism** [comment]
@cwharris please don't force push to open PR branches. It can mess up existing comments on the PR. It's OK to rebase and force push to branches that are not yet part of an open PR, but once they are public it can lead to problems.

File: cpp/include/cudf/copying.hpp [hunk 3]
```diff
@@ -26,6 +26,12 @@ typedef struct CUstream_st* cudaStream_t;
 
 namespace cudf {
 
+enum mask_allocation_policy {
```
Comment (harrism) [inline cpp/include/cudf/copying.hpp]: The enum itself still needs doxygen block. There's an example here: https://docs.google.com/document/d/10Ff4VZlSsizHdF5r9Zrw7dIwlPD0yWWSmNJ6qter7r8/edit#heading=h.rgm5sgyxoiv6

File: cpp/include/cudf/copying.hpp [hunk 4]
```diff
@@ -26,6 +26,12 @@ typedef struct CUstream_st* cudaStream_t;
 
 namespace cudf {
 
+enum MaskAlloc {
```
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/include/cudf/copying.hpp]: Notice you can no longer see this comment in the code view for this PR. That's because you force pushed. :)

**Comment by cwharris** [comment]
It looks like the merge is blocked due to a 10.1 failure. I thought 10.1 wasn’t suppose to be blocking merges?

**Comment by harrism** [comment]
Hmmm, CI is failing...

**Comment by cwharris** [comment]
@harrism #2809 should fix CI, I'm told.

**Pull request merged**